code/API_definitions/qod-api.yaml

```diff
@@ -430,6 +444,44 @@ components:
           - address - an IPv4 number in dotted-quad form 1.2.3.4. Only this exact IP number will match the flow control rule.
           - address/mask - an IP number as above with a mask width of the form 1.2.3.4/24.
             In this case, all IP numbers from 1.2.3.0 to 1.2.3.255 will match. The bit width MUST be valid for the IP version.
+    PublicIpv4Addr:
+      description: |
+        Public (routable) source IP address and public port seen by, and identifying the device to, the application server
+      type: object
+      properties:
+        public_address:
+          $ref: "#/components/schemas/SingleIpv4Addr"
+        public_port:
+          $ref: "#/components/schemas/Port"
+      required:
+        - public_address
+        - public_port
+      example:
+        {
+          "public_address": "85.255.235.230",
+          "public_port": 60276
+        }
+    PrivateIpv4Addr:
+      description: |
+        Private (non-routable) source IP address directly allocated to the device, and the public (routable) source IP seen by and identifying the device to the application server. If the device does not have a public IP address, then the public_address value should be specified as the private address as well to indicate this.
```

Comment:
If the device does not have a ~~public IP address~~`private IP address`, then...